### DISCUSSION TITLE
How to set and access a Workflow variable?

### DISCUSSION THREAD

**Discussion by cwiechmann** [community/community#25295] [Actions]
<p>Hi,</p>
<p>I would like to simplify my workflow by using a variable at certain places. That workflow ultimately becomes some kind of a template and I would like to make it as easiest as possible for people to maintain it, when they have created their own copy of that workflow. Therefore I was thinking to declare a global variable and use it in the workflow at many places:</p>
<p>It basically looks like this and people should only be required to maintain the pluginId:</p>
<pre><code>env:
  pluginId: 'plugin-fn-xml-node'

on:
  push:
    paths:
      - ${{env.pluginId}}/**
      - .github/workflows/**
  pull_request:
    paths:
      - ${{env.pluginId}}/**
      - '.github/workflows/**'

jobs:
  build:
    env:
      working-directory: ${{env.pluginId}}
    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [8.x, 10.x, 12.x]

    steps:
.....
</code></pre>
<p>That fails with the error message:</p>
<pre><code>Your workflow file was invalid: The pipeline is not valid. .github/workflows/fn-xml-node-tests.yml (Line: 7, Col: 9): Unrecognized named-value: 'env'. Located at position 1 within expression: env.pluginId,.github/workflows/fn-xml-node-tests.yml (Line: 11, Col: 9): Unrecognized named-value: 'env'. Located at position 1 within expression: env.pluginId
</code></pre>
<p>It’s stated here: <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#env" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#env</a> that workflow (global) variables are supported, but I don’t find a way to get it working. Unfortunately the documentation isn’t much talking about Workflow-Variables.</p>
<p>Can someone tell me, if this is possible? And if yes, what I’m doing wrong.</p>
<p>Thanks!</p>
<p>Chris</p>

**Comment by rigobcastro** [comment]
<p>I guess you don’t need the env prefix,  just use the var name {{ pluginId }}</p>

**Comment by BrightRan** [comment]
<p>Currently, we have no ways to call environment variables on job level, ideally the syntax <strong>${{ env.VAR_NAME }}</strong> should work but just has not been implemented yet. It’s being tracked internally, but I don’t have any timeline to share yet. On step level, <strong>${{ env.VAR_NAME }}</strong> is working fine.</p>
<p>In addition, when defining environment variables, you only can use <strong>alphanumeric</strong> characters and <strong>_</strong>  character (not - character) to name the environment variables, and environment variables are case-sensitive.</p>

**Comment by cwiechmann** [comment]
<p>Hi <a class="mention" href="/rigobcastro">@rigobcastro</a>,</p>
<p>Thanks for the response. Unfortunately the error also that way remains. I tried:</p>
<p>${{ pluginId }} and {{ pluginId }}</p>
<p>Any other ideas?</p>

**Comment by cwiechmann** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a>,</p>
<p>That’s interesting feedback. But then I don’t understand why I can even declare an environment variable at workflow level. If I declare something, it should be possible to reference it somewhere in the workflow file.<br>
Also, it’s stated here that the feature is available: <a href="https://github.blog/changelog/2019-10-01-github-actions-new-workflow-syntax-features/" rel="nofollow noopener">https://github.blog/changelog/2019-10-01-github-actions-new-workflow-syntax-features/</a></p>
<p>Or is this something different? And really thanks for the hint regarding the allowed characters.</p>

**Comment by BrightRan** [comment, marked as answer]
<p>Yeah, we can define environment variables at workflow level, job level and step level.<br>
And ideally, the workflow level environment variables can be used at all jobs and steps in the workflow, the job level environment variables can only be used at all steps in the current job, and the step level environment variables can only be used at current step.</p>
<p>But now, the workflow level environment variables can’t be used at job level. I’m not sure why it is not supported initially. This problem had been reported to the appropriate engineering team for further investigation and evaluation.</p>
<p>As I metioned previously, the engineering team is considering to make the syntax  <strong>${{ env.VAR_NAME }}</strong>  also work at job level, but they do not share any timeline.</p>
<p>Apologize for the inconvenience caused to you.</p>

**Reply by cowwoc** [reply]
This seems to be working as of 2023/03/07

**Reply by dkrasnikov-caret** [reply]
No it is not. Still getting error env is not available on job level

**Comment by ohtangza** [comment]
<p>How can we know when this is available? We want to re-use a global scope variable in more than 5 jobs.</p>
<p>env:<br>
  CI_ACTIVATED: &gt; # <a href="https://stackoverflow.com/a/53099054/1609620" rel="nofollow noopener">https://stackoverflow.com/a/53099054/1609620</a><br>
    (github.event.action == ‘synchronize’ &amp;&amp; contains(github.event.pull_request.labels.*.id, 584780719))<br>
    ||<br>
    (github.event.action == ‘labeled’ &amp;&amp; github.event.label.id == 584780719)</p>

**Comment by bfleming-ciena** [comment]
<p>Use a secret?</p>

**Comment by stefanmaric** [comment]
<p>Three months have passed and it still seems a problem. Proper scoping and cascading of env vars for the workflow definiton seems to me like a pretty important feature.</p>

**Comment by cerna** [comment]
<p>Hi,</p>
<p>I am not completely sure what you mean, but the following currently works:</p>
<pre><code>name: Show env

on:
  push:
    branches:
      - '*'
env:
  somevar: 'lastvar'
jobs:
  show:
    runs-on: ubuntu-latest
    steps:
      - name: Is variable exported?
        run: |
          echo "${{ env.somevar }}"
</code></pre>
<p>What does not work is to change this variable from one <strong>job</strong> and have it available in another <strong>job</strong> (of course with <strong>needs</strong> ). To use it for settings sharing. (What I do now is to create a JSON at the end of settings job and then use it in all following jobs.)</p>

**Comment by m3hari** [comment]
<p>+1 </p>
<p>This is an important feature!</p>
<p>Perhaps the docs should be very clear about this if is not ready yet. I spent hours without realizing it is not supported.</p>
<p>Thanks</p>

**Comment by meiswjn** [comment]
<p>Any updates on this? This is a crucial feature.</p>

**Comment by bjmd** [comment]
<p>I’m not sure if its a recent thing, but the functionality works fine for me. Global env which is usable in steps.</p>

**Comment by jasonkarns** [comment]
<p>The problem states that the <em>workflow</em> env var is not available at <em>job</em> level configuration.</p>
<p>The example you demonstrate here (which does work) is to use a workflow level env var at the <em>step</em> level. If you change your example to move the <code>${{ env.somevar }}</code> reference from the step up to the job level, it will fail.</p>

**Comment by jasonkarns** [comment]
<p>The problem clearly states the workflow env is not usable at the <em>job</em> level. Step level references are fine, but job-level references break (as demonstrated in the example).</p>

**Comment by meiswjn** [comment]
<p>Probably simplest workaround right now:</p>
<ol>
<li>Set Environment Var</li>
<li>Step takes it as input and sets it as the step’s output.</li>
<li>Environment variable can now be accessed via ‘needs’.</li>
</ol>
<p>Of course, this lacks quite a few characteristics of typical env-vars, most importantly the ability to change them dynamically in a run.</p>

**Comment by softmarshmallow** [comment]
<p>Any updates on this? my usecase is creating a firebase.config.js file via <code>echo ${{SOME_SECRET_CONTENT}}</code></p>

**Comment by powerbuoy** [comment]
<p>I would also like to see this implemented. I’m setting a couple of env vars on the workflow level and then I’m trying to use them on the job-level’s if-statement;</p>
<pre><code class="lang-auto">name: Deploy
on: push

env:
  SRC: './'
  PROD: 'user@prod-host:/path/to/site/'
  # STAGE: 'user@stage-host:/path/to/site/'

jobs:
  deploy:
    # Only run job if stage is set and this is NOT a version tag push OR if prod is set and this IS a version tag push
    if: (env.STAGE &amp;&amp; !startsWith(github.ref, 'refs/tags/v')) || (env.PROD &amp;&amp; startsWith(github.ref, 'refs/tags/v'))
</code></pre>

**Comment by hoxu** [comment]
<p><a class="mention" href="/brightran">@brightran</a>: any update on <code>on.*.paths</code> getting support for env variables?</p>

**Comment by dezren39** [comment]
<p>+1, this feature is essential</p>

**Comment by PiotrPopiolekKW** [comment]
Is it still on the table to enable this feature?

**Comment by bar-maiost-juno** [comment]
This is a basic and essential feature, is it on the roadmap?

**Comment by cjbottaro** [comment]
Almost three years later and it still doesn't work. I love using Azure Pipelines with a paint job.

**Comment by vanduc2708** [comment]
15/12/2024 it still doesn't work for job level

**Comment by rogerkeithi** [comment]
02/05/2025 it still doesn't work for job level
